### PR TITLE
adding load_text, load_image, load_dict to fluent.py

### LINKED ISSUE #6428: [FR] Add APIs for loading text, JSON, and images to mlflow.artifacts
### Willingness to contribute

No. I cannot contribute this feature at this time.

## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers. We’ve identified this feature as a highly requested addition to the MLflow package based on community feedback. 
We're seeking a community contribution for the implementation of this feature and will enthusiastically support the development and review of a submitted PR for this.

## Feature scope

This roadmap feature’s complexity is classified as:

- [X] `good-first-issue`: This feature is limited in complexity and effort required to implement.
- [ ] `simple`: This feature does not require a large amount of effort to implement and / or is clear enough to not need a design discussion with maintainers.
- [ ] `involved`: This feature will require a substantial amount of development effort but does not require an agreed-upon design from the maintainers. The feedback given during the PR phase may be involved and necessitate multiple iterations before approval. (Please bear with us as we collaborate with you to make a great contribution)
- [ ] `design-recommended`: This is a substantial feature that should have a design document approved prior to working on an implementation (to save your time, not ours). After agreeing to work on this feature, a maintainer will be assigned to support you throughout the development process.

## Proposal Summary

Introduce the following APIs to MLflow:

1. `mlflow.artifacts.load_text()`: Takes a URI referring to a text file as a parameter, downloads the file, and loads the file's contents as a string that is returned to the caller.

2. `mlflow.artifacts.load_json()`: Takes a URI referring to a JSON file as a parameter, downloads the file, and loads the file's contents as a dictionary that is returned to the caller via JSON parsing, throwing if the file cannot be parsed as JSON.

3. `mlflow.artifacts.load_image()`: Takes a URI referring to an image file as a parameter, downloads the file, and loads the file's contents as an image using Pillow's `Image.load()` function (https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.load), throwing if the file cannot be loaded as an image. The Pillow image object should be returned to the caller; if the function is called in a Jupyter notebook environment, it should also be displayed automatically.

### Motivation

> #### What is the use case for this feature?

Provide users a way to easily load common MLflow artifacts as Python objects.

> #### Why is this use case valuable to support for MLflow users in general?

Many users log text, JSON, and images, and they need to load them later for further analysis / processing.

> #### Why is it currently difficult to achieve this use case?

MLflow's `mlflow.artifacts.download_artifacts()` API (https://mlflow.org/docs/latest/python_api/mlflow.artifacts.html#mlflow.artifacts.download_artifacts) only downloads file contents to disk; it is up to users to parse them.

### Details

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by subramaniam02**
## Related Issues/PRs

closes #6428

## What changes are proposed in this pull request?

adding 3 methods to fluent.py
- load_text
- load_dict
- load_image

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

https://github.com/mlflow/mlflow/issues/6428

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/__init__.py
Comment (harupy) [inline mlflow/__init__.py:172]: Are we sure this is correct? The proposal in https://github.com/mlflow/mlflow/issues/6428 is to add `mlflow.artifacts.load_text()`, `mlflow.artifacts.load_json()`, and `mlflow.artifacts.load_image()`.

Comment (subramaniam02) [reply mlflow/__init__.py]: Sorry, my bad. I started right but ended in different direction. Have made the changes accordingly ! Thanks for correction :)

**Comment by subramaniam02** [comment]
I am going to away for next couple of days, expect less communication from me :(

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:94]: Do we need this try-catch?

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:88]:
```suggestion
    with tempfile.TemporaryDicrectory() as tmpdir:
        local_artifact = download_artifacts(artifact_uri, dst_path=tmpdir)
        with open(local_artifact) as local_artifact_fd:
            ...
```

Can we use `TemporaryDicrectory` to immediately remove the downloaed file after reading its content?

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:159]: We should check the PIL availability first. 

```suggestion
    try:
        from PIL import Image
    except ImportError as exc:
        raise ImportError(
            "`load_image` requires Pillow. Please install it via: pip install Pillow"
        ) from exc
    local_artifact = download_artifacts(artifact_uri)
```

File: mlflow/artifacts/__init__.py
Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: good point ! Have modified code accordingly :)

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:129]:
```suggestion
```

```suggestion
```

Let's remove this. `raise MlflowException` shows a traceback.

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:170]: Do users **_always_** want to show the loaded image? Maybe not.

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:173]:
```suggestion
```

File: mlflow/artifacts/__init__.py
Comment (harupy) [reply mlflow/artifacts/__init__.py:170]: If this function is called in a loop, an image is displayed in each iteration, which might be annoying.

Comment (harupy) [reply mlflow/artifacts/__init__.py:170]: Wait, do we even need to call `IPython.display`?

image

Comment (harupy) [reply mlflow/artifacts/__init__.py:170]: Test on Jupyter:

image

Comment (harupy) [reply mlflow/artifacts/__init__.py:170]: Test on Databricks

image

Comment (harupy) [reply mlflow/artifacts/__init__.py:170]:
```suggestion
```

Can we remove this block?

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:112]:
```suggestion
            mlflow.log_dict({"mlflow-version": "0.28", "n_cores": "10"}, "config.json")
```

We can use `log_dict`.

File: mlflow/artifacts/__init__.py
Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: I thought it is good to have. What if user tries to use the `load_text` method with an artifact file that contains image ?

```
with mlflow.start_run() as run:
    image = Image.new("RGB", (100, 100))
    artifact_uri = run.info.artifact_uri
    mlflow.log_image(image, "image.png")
    artifact_uri = run.info.artifact_uri
    file_content = mlflow.artifacts.load_text(artifact_uri + "/image.png")
    print(file_content)
```

Error traceback
```
Traceback (most recent call last):
  File "/Users/subras/github/mlflow-fork/mlflow/artifacts/__init__.py", line 93, in load_text
    return str(local_artifact_fd.read())
  File "/Users/subras/.pyenv/versions/3.7.13/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x89 in position 0: invalid start byte

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 47, in <module>
    file_content = mlflow.artifacts.load_text(artifact_uri + "/image.png")
  File "/Users/subras/github/mlflow-fork/mlflow/artifacts/__init__.py", line 95, in load_text
    raise MlflowException("Unable to form a str object from file content", BAD_REQUEST)
mlflow.exceptions.MlflowException: Unable to form a str object from file content
```

File: mlflow/artifacts/__init__.py
Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: @harupy  I tested below code block in notebook. When user calls load_image, image does not get displayed. When `Image.new` is called, yes notebook displays the image but not for `Image.open` which is used in load_image method. according to description of FR, image should be displayed when executed in notebook. I understand your concerns regarding whatif called in a loop, etc. What do you think of changing the function signature to`def load_image(artifact_uri: str, display_image=False)`

```
import mlflow
from PIL import Image
with mlflow.start_run() as run:
    image = Image.new("RGB", (100, 100))
    artifact_uri = run.info.artifact_uri
    mlflow.log_image(image, "image.png")
    image  = mlflow.artifacts.load_image(artifact_uri + "/image.png")
```

Comment (harupy) [reply mlflow/artifacts/__init__.py]: It's because you assigned the result of `mlflow.artifacts.load_image` to a variable.

image

Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: @harupy  Here is the behavior I see, inside a with context manager and one without it. IF users dont call `load_image` within the `with mlflow.start_run()` context, then you are right, they is no need for explicit `IPython.display()`. 

Screen Shot 2022-08-25 at 10 06 35 PM

Comment (harupy) [reply mlflow/artifacts/__init__.py]: I think you can:

```pytohn
with mlflow.start_run() as run:
    ...
    image  = mlflow.artifacts.load_image(artifact_uri + "/image.png")

image
```

Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: @harupy yes, above suggestion works !! From my side I have pushed PR review changes, please let me know if there is any more details to be taken care of before the merge :)

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:160]:
```suggestion
            image_obj = Image.open(local_artifact)
            image_obj.load()
```

https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.open:

> This is a lazy operation; this function identifies the file, but the file remains open and the actual image data is not read from the file until you try to process the data (or call the [load()](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.load) method).

File: tests/artifacts/test_artifacts.py
Comment (harupy) [inline tests/artifacts/test_artifacts.py:80]: Can we use `namedtuple` instead of plain `tuple`?

File: mlflow/artifacts/__init__.py
Comment (subramaniam02) [inline mlflow/artifacts/__init__.py:126]:
```suggestion
def load_image(artifact_uri: str) -> 'PIL.Image':
```

@harupy  What is your suggestion on this change ?

File: tests/artifacts/test_artifacts.py
Comment (harupy) [reply tests/artifacts/test_artifacts.py:80]: Example:

```python
Artifact = namedtuple("Artifact", ["uri", "content"])

@pytest.fixture()
def run_with_text_artifact():
    artifact_path = "test/file.txt"
    artifact_content = "This is a sentence"
    with mlflow.start_run() as run:
        mlflow.log_text(artifact_content, artifact_path)

    artifact_uri = str(pathlib.PurePosixPath(run.info.artifact_uri) / artifact_path)
    return Artifact(artifact_uri, content)


def test_load_text(run_with_text_artifact):
    artifact = run_with_text_artifact
    assert mlflow.artifacts.load_text(artifact.uri) == artifact.content
```

File: mlflow/artifacts/__init__.py
Comment (harupy) [reply mlflow/artifacts/__init__.py:126]: I think this is useless because Pillow is not typed:

https://pillow.readthedocs.io/en/stable/_modules/PIL/Image.html#open

Comment (harupy) [reply mlflow/artifacts/__init__.py:126]:
```python
# a.py

from PIL import Image

reveal_type(Image.open("foo"))
```

Output:

```
> mypy a.py
a.py:1: error: Skipping analyzing "PIL": module is installed, but missing library stubs or py.typed marker
a.py:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
a.py:3: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
```

Comment (harupy) [reply mlflow/artifacts/__init__.py:126]: We need to do something like this to make type hint work:

```
from PIL import Image

import typing

if typing.TYPE_CHECKING:
    from PIL import Image


def get_image() -> "Image.Image":
    from PIL import Image

    return Image.open("test.png")


img = get_image()

```

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:66]:
```suggestion
    Loads the artifact contents as a string.
```

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:97]:
```suggestion
    Loads the artifact contents as a dictionary.
```

File: mlflow/artifacts/__init__.py
Comment (harupy) [inline mlflow/artifacts/__init__.py:128]:
```suggestion
    Loads the artifact contents as a ``PIL.Image.Image`` object.
```

**Review by harupy — APPROVED** [review]
LGTM once the remaining comments are addressed. @dbczumar Could you also take a look?

File: tests/artifacts/test_artifacts.py
Comment (subramaniam02) [reply tests/artifacts/test_artifacts.py:80]: Great suggestion !!

**Review by dbczumar — APPROVED** [review]
LGTM! Awesome work, @subramaniam02 !

Excited to merge once the small docs issue is addressed: https://github.com/mlflow/mlflow/pull/6475/files#r957996633

File: mlflow/artifacts/__init__.py
Comment (dbczumar) [inline mlflow/artifacts/__init__.py:126]: I think this typehint is causing a docstring error:

```
/home/circleci/project/mlflow/artifacts/__init__.py:docstring of mlflow.artifacts.load_image:: WARNING: py:class reference target not found: Image.Image
```

I think it's acceptable to remove the type hint for now.

Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: @dbczumar Sorry about missing CI failures. I have removed the type hint. Will look into this in detail and create enhancement PR.

Comment (subramaniam02) [reply mlflow/artifacts/__init__.py]: @dbczumar done :)

**Comment by dbczumar** [comment]
Merging! Thanks so much, @subramaniam02 !

**Pull request merged**
